### ISSUE TITLE
Inconsistent output for `what_sites` and `get_info`

### ISSUE THREAD

**Issue by SorooshMani-NOAA** [USGS-python/dataretrieval#78]
When the input parameters of `what_sites` and `get_info` results in empty data set, the return value is a dataframe with different columns names. Instead one would expect an empty dataframe with the same columns as the normal output.

Example:
```python
nwis.what_sites(**{'stateCd': 'ny', 'parameterCd': ['00056', '00058', '00059', '50051', '74072'], 'hasDataType': 'iv'})
```
results in (correctly)
```
(  agency_cd          site_no  station_nm site_tp_cd  dec_lat_va  dec_long_va coord_acy_cd dec_coord_datum_cd  alt_va  alt_acy_va alt_datum_cd   huc_cd
 0      USGS  405412072251201  S133926. 1         GW   40.903333   -72.420000            S              NAD83   95.15         5.0       NAVD88  2030202
 1      USGS  425227073212401      RE 108         GW   42.874375   -73.356731            H              NAD83  451.00         4.3       NAVD88  2020003
 2      USGS  425237073210903      RE 293         GW   42.877197   -73.352694            H              NAD83  430.20         1.0       NAVD88  2020003,
 <dataretrieval.utils.Metadata at 0x2b06b7d7ba60>)
```
while 
```python
nwis.what_sites(**{'stateCd': 'fl', 'parameterCd': ['00056', '00058', '00059', '50051', '74072'], 'hasDataType': 'iv'})
```
results in
```
(Empty DataFrame
 Columns: [<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>Error report</title><style type="text/css"><!--H1 {font-family:TahomaArialsans-serif;color:white;background-color:#525D76;font-size:22px;} H2 {font-family:TahomaArialsans-serif;color:white;background-color:#525D76;font-size:16px;} H3 {font-family:TahomaArialsans-serif;color:white;background-color:#525D76;font-size:14px;} BODY {font-family:TahomaArialsans-serif;color:black;background-color:white;} B {font-family:TahomaArialsans-serif;color:white;background-color:#525D76;} P {font-family:TahomaArialsans-serif;background:white;color:black;font-size:12px;}A {color : black;}HR {color : #525D76;}--></style> </head><body><h1>HTTP Status 404 - No sites found matching this request server=[sdas01]</h1><hr/><p><b>type</b> Status report</p><p><b>message</b>No sites found matching this request server=[sdas01]</p><p><b>description</b>The requested resource is not available.</p><hr/><h3>Error Report</h3></body></html>]
 Index: [],
 <dataretrieval.utils.Metadata at 0x2b06b7d7f730>)
```

**Comment by elbeejay** [comment]
@SorooshMani-NOAA thank you for flagging this and providing a nice functional example of the issue.

While I understand your desire for an empty dataframe with matching columns, that is not likely to be possible given how `dataretrieval` functions. The query resulting in an empty dataset that you've provided as an example below, is actually an HTTP 404 error that `dataretrieval` is not handling correctly and is trying to coerce into a dataframe.

The approach consistent with the way `dataretrieval` handles other HTTP errors, would be to return a `ValueError` and some message about the empty query (e.g., our handling of 414 errors: https://github.com/USGS-python/dataretrieval/pull/65).

Practically, `dataretrieval` does not have a list of the expected column headers, so I do not think it is going to be possible to return an empty dataframe with the same columns. However I would like to ask: **Would an error disrupt workflows in this case? Would it be preferable to return a truly empty dataframe (`pd.DataFrame()`) and a warning?** We could more generally switch `dataretrieval` HTTP error handling to warnings + empty dataframes if that is more user-friendly (@thodson-usgs would want you to weigh in on this too).

**Comment by SorooshMani-NOAA** [comment]
@elbeejay,  thank you for the detailed explanation. I would prefer having exceptions and errors. Having an error does require me to change my code, but that's minimal. Treating this as a special case by returning an empty dataframe is just going to open the door to other special cases, which just make everything less transparent and harder to maintain on your side. Which in turn hurts everyone using it (including us) in the long run.

So in short, it'd be better if you do what's consistent with the rest of the errors/exceptions and I'll adapt.

**Comment by thodson-usgs** [comment]
Maybe I'm old fashion, but I like the behavior that an empty query returns and empty dataframe (perhaps along with a warning).
That's how SQL would do it, and a common design pattern with `dataretrieval` is making many small queries then merge them together, which is easy with empty dataframes (no rows, no columns: `pd.DataFrame()`). I agree we shouldn't chase every special case, but I've found it to be a feature more than a bug.
I think something like:
```python
Empty DataFrame
Columns: []
Index: []
Empty DataFrame
Columns: []
Index: []
Warning: No data found matching request: ... blah blah metadata...
```

**Comment by elbeejay** [comment]
I agree with the "special case" thinking - I can plan to switch *all* of the HTTP error handling (so beyond just the 404 error) to a warning, and adjust some of the file reading functions (e.g., `_read_rdb`) to return an empty data frame in those cases. This would represent a departure from existing behavior unfortunately.

A middle ground to preserve existing behavior and still throw errors would be to add a `strict` parameter as a boolean that would throw an error when `True`, or a warning when `False`. But not sure that complexity is warranted as it'd either become part of `**kwargs` or would be a key word parameter added to a ton of functions...

**Comment by thodson-usgs** [comment]
I've confirmed that different USGS services behave differently. For example,
`https://waterservices.usgs.gov/nwis/dv?format=rdb&startDT=1800-01-01&endDT=1801-01-01&sites=03339000`
returns an empty data object rather than a 404.
I agree that isn't ideal though it isn't exactly within our scope either: the bug is in the REST server, not our Python wrapper.
@elbeejay and I will have to stew on this, I think. 

@ldecicco-usgs, how would you handle this in the R package?

**Comment by elbeejay** [comment]
Going to do some pinging so we don't all lose sight of this issue - @thodson-usgs @ldecicco-USGS

**Comment by ldecicco-USGS** [comment]
Ideally it would be great to return an empty data.frame (in R... whatever you call it in Python) with the correct columns. In the R package, we do that for _some_ of the functions. So, the "hand-holding" functions like "readNWISdv" "readNWISuv"...there is a check for no data, and if that happens it comes back with an empty data.frame that's hard-coded.

Ideally....the services would come back and say "yes you made a valid query" and "here's the columns we _would_ return if there was data"....but I don't think any of the NWIS services do that (off hand? don't quote me on that).

The empty return in this FL example is a good example of why it's hard. The http header that is returned is a 404. There's a useful message that is returned if you go to the URL:
https://waterservices.usgs.gov/nwis/site/?stateCd=FL&parameterCd=00056,00058,00059,50051,74072&hasDataType=iv&format=mapper
But other services tell us that same "No sites found matching this request" with a 200 header (instead of the 404). 

Anyway, I guess 3 options? 
1. Someone go through all the functions we do, get the format that comes back with data and documents that so we can hard-code all the functions that we are confident will only come back with that format.
2. See if we can get the web services to return with an empty table (or whatever) if the query was valid but there was no data.
3. Leave the no-data yet valid results to be at the whim of what NWIS/WQP gives us. (maybe pepper your documentation on how to add a "try-fail" type check if you are trying to loop through states or something like that).


R's dataRetrieval does a mis-mash of 1 and 3. 1 for some functions, 3 for others. I'm happy to try to increase the number of 1's if someone wanted to compile that info. 

I've hard coded _some_ of the functions to come back with an empty data frame if I'm very confident on the data structure. But I don't do that when there are possibilities that I don't know the data format.

**Comment by elbeejay** [comment]
Thanks for that input @ldecicco-USGS - seems like there is no "silver bullet" for this problem (short of upstream fixes to the web services).

My two cents are that your option 1 is not very appealing due to both the work required, but also the fragility with respect to any future changes to the services. 2 happens for some of the services by default, but not all of them. 3 feels like the best solution for a thin wrapper, basically let the query fail if a 404 error is returned (but some services might pass along an empty data frame which would be let through if the header is a 200 code).

**Comment by elbeejay** [comment]
@thodson-usgs the more I think about it, the more I think the solution in #79 is the most maintainable from our perspective (handling HTTP errors as they are given). It isn't ideal, but if upstream webservices change, or are fixed to all serve empty responses for valid queries, we won't have to make any changes here.

**Comment by thodson-usgs** [comment]
Agreed. I'll review #79 but leave this issue open. I'll ask them to fi this upstream.

**Comment by thodson-usgs** [comment]
Sent to NWIS:

Good day.
I'm writing on behalf of the USGS developers that maintain the R and Python dataretrieval libraries that wrap the NWIS REST Server. Some of the NWIS services have inconsistent behavior. We considered downstream fixes but ultimately decided they would be fragile and that the best solution was to fix the problem upstream.

Here is the description of the bug:
Query 1 returns valid data:
https://waterservices.usgs.gov/nwis/site?stateCd=ny&parameterCd=00056%2C00058%2C00059%2C50051%2C74072&hasDataType=iv&format=rdb'

Query 2 is correctly formatted but returns 404 because no data was found.
https://waterservices.usgs.gov/nwis/site?stateCd=fl&parameterCd=00056%2C00058%2C00059%2C50051%2C74072&hasDataType=iv&format=rdb

An incorrectly formatted query for any service returns a 400 error.

However other NWIS services have different behavior.
Query 3 returns an empty table when no data were found, rather than a 404 
https://waterservices.usgs.gov/nwis/dv?format=rdb&startDT=1800-01-01&endDT=1801-01-01&sites=03339000

Query 2 and Query 3 should have the same behavior. For Query 2, could the server respond with an empty table or at least a HTTP status 200, rather than a 404?

Thank you for your assistance.

**Comment by thodson-usgs** [comment]
"Thanks for the note regarding the [REST Web Services (usgs.gov)](https://waterservices.usgs.gov/rest/). There are certainly a number of improvements (including your great suggestion) that could be made to the current implementation of the water services. I think at this time the code is more or less frozen though, as developers have begun to modernize the services."
